app/organism_loop.py

```
@@ -0,0 +1,280 @@
+    matched_constraint_id: str | None
+    state_root: str
+    evidence: Mapping[str, Any]
+
+
+def _function_fingerprint(fn: Any) -> str:
+    try:
+        source = inspect.getsource(fn)
+    except (OSError, TypeError):
+        code = getattr(fn, "__code__", None)
+        if code is None:
+            raise ValueError(f"cannot fingerprint function {fn!r}")
```

Ruff (0.16.1)

[warning] Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling

(B904)